drivers: firmware: scmi: add more protocols and APIs

yangbolu1991 opened the pull request:
This patch-set is to add more SCMI clock protocol APIs and power domain protocol driver.
These are essential for platform to configure clock and control power domain.
These had been verified on NXP i.MX95 platforms.

Thank you very much.

yangbolu1991 committed:
drivers: firmware: scmi: add power domain protocol

Added helpers for ARM SCMI power dmomain protocol.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
@LaurentiuM1234 Thanks. 
Logically, this is an independent change to add more APIs/protocol for SCMI.
As we know, setting clock parent and power domain control are not supported in SCMI, which should be basic features.

Actually I will send NETC support PR, which needs this to set clock and power up NETCMIX domain. That will include,
- new i.mx netc drivers
- imx95 platform netc support.

So, not sure if the PR will be huge making all changes together with this in one PR.
But I'm ok for either.

THanks.

LaurentiuM1234 commented:
> Below is just FYI. We also worked on Zephyr SCMI implementation on raspberry pi 5 [1], but, of course, now we will migrate on already existing solution (Our solution implements SMC/HVC transport only). But what you might find interesting/useful is that we added SHELL interface and SCMI sample application which allows to test SCMI - at least manually.
> 
> [1] [xen-troops#124](https://github.com/xen-troops/zephyr/pull/124)

great to see more platforms using the SCMI support! Do you plan on also upstreaming the patches to Zephyr? I think it would be good to have another upstream platform using SCMI so we could have more coverage. Also, IMO, the sample is a very good idea since we're lacking in that department.

GrygiriiS commented:
@LaurentiuM1234 Hope to have smth this week. I was able to port on top of current code, need to look at it one more time. Current state in https://github.com/GrygiriiS/zephyr/commits/zm_rpi5_dev_scmi/

yangbolu1991 committed:
drivers: firmware: scmi: add more APIs of clock management protocol

Added more APIs for ARM SCMI clock management protocol.
- scmi_clock_rate_set
- scmi_clock_parent_get
- scmi_clock_parent_set

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

LaurentiuM1234 commented on drivers/firmware/scmi/power.c:
```diff
@@ -0,0 +1,90 @@
+	reply.len = sizeof(status);
+	reply.content = &status;
+
+	ret = scmi_send_message(proto, &msg, &reply);
```
This function is synchronous so if someone sets the ASYNC bit it will come in contradiction with it. For now, since we don't support ASYNC messages, maybe we can return an error if someone attempts to call this function with that bit set? Or perhaps force it to 0 and print a warning?

yangbolu1991 replied:
The scmi driver should be standard. I think it's not good to limit only SYNC flag here.
Actually, if ASYNC is used, I think the reply from system manager will be error, right?

LaurentiuM1234 replied:
we're limited to SYNC because that's the only option the support allows. Allowing people to set the ASYNC flag and still using SYNC messaging behind the scenes is not the way IMO. This deserves a comment at least.

LaurentiuM1234 commented on drivers/firmware/scmi/power.c:
```diff
@@ -0,0 +1,90 @@
+#include <zephyr/drivers/firmware/scmi/power.h>
+#include <string.h>
+
+DT_SCMI_PROTOCOL_DEFINE_NODEV(DT_INST(0, arm_scmi_power), NULL);
```
A note on this: not exactly ideal to put this here as we'd normally have a generic power domain driver using this, but since Zephyr doesn't really support DT nodes providing multiple PDs it becomes a bit tricky. For now, I'd say it's fine to go with this.

yangbolu1991 replied:
Okay, thanks:)

LaurentiuM1234 reviewed (commented):
Sorry for the (very) long review time. Some comments below. Also, is there a particular reason you didn't implement the `clock_control_set_rate()` function? Should be useful for https://github.com/zephyrproject-rtos/zephyr/pull/81113.

yangbolu1991 commented:
> Sorry for the (very) long review time. Some comments below. Also, is there a particular reason you didn't implement the `clock_control_set_rate()` function? Should be useful for #81113.

I think the common clock control driver is missing function like set_parent.
It's not able to directly use set_rate since we have to set parent clock before that.
So, to quick support NXP platform, I use scmi APIs directly.
Thanks.

yangbolu1991 commented on drivers/firmware/scmi/clk.c:
```diff
@@ -61,6 +66,113 @@ int scmi_clock_rate_get(struct scmi_protocol *proto,
+	return 0;
+}
+
+int scmi_clock_parent_set(struct scmi_protocol *proto, struct scmi_clock_parent_config *cfg)
```
Agree, I can't understand why I encased the parameters too.
Thanks.

LaurentiuM1234 commented:
BTW, could you please also add the PD protocol to the list of supported protocols in https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/hardware/arch/arm-scmi.rst? Thanks!

yangbolu1991 commented:
Updated to v3. Changes include,

- Returned not supported for ASYNC flag for clk/power API. I had some misunderstand before.
- Converted to use clk_id/parent_id as scmi_clock_parent_set() arguments.
- Added power domain protocol in zephyr scmi doc.

Thanks a lot.

LaurentiuM1234 reviewed (approved):
good with me @dleach02

mmahadevan108 commented:
@LaurentiuM1234 , would you like to add a change PR to make yourself the maintainer for ARM SCMI?
